Repository: plone.app.layout


Branch: refs/heads/2.3.x
Date: 2017-01-13T14:58:42+01:00
Author: Maurits van Rees (mauritsvanrees) <maurits@vanrees.org>
Commit: https://github.com/plone/plone.app.layout/commit/15156a4e7970bce7cf47351c29dab63dad0d1eaa

Fix error in viewlet when related dexterity item has been deleted.

Files changed:
M CHANGES.rst
M plone/app/layout/viewlets/content.py
M plone/app/layout/viewlets/tests/test_content.py

diff --git a/CHANGES.rst b/CHANGES.rst
index a53be63..52f8f9b 100644
--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -14,7 +14,8 @@ New features:
 
 Bug fixes:
 
-- *add item here*
+- Fix error in viewlet when related dexterity item has been deleted.
+  [maurits]
 
 
 2.3.16 (2016-12-19)
diff --git a/plone/app/layout/viewlets/content.py b/plone/app/layout/viewlets/content.py
index 83b8c92..7892997 100644
--- a/plone/app/layout/viewlets/content.py
+++ b/plone/app/layout/viewlets/content.py
@@ -176,6 +176,10 @@ def related2brains(self, related):
         brains = []
         for r in related:
             path = r.to_path
+            if path is None:
+                # Item was deleted.  The related item should have been cleaned
+                # up, but apparently this does not happen.
+                continue
             # the query will return an empty list if the user has no
             # permission to see the target object
             brains.extend(catalog(path=dict(query=path, depth=0)))
diff --git a/plone/app/layout/viewlets/tests/test_content.py b/plone/app/layout/viewlets/tests/test_content.py
index 3bf71bc..04b3f58 100644
--- a/plone/app/layout/viewlets/tests/test_content.py
+++ b/plone/app/layout/viewlets/tests/test_content.py
@@ -106,6 +106,15 @@ def testRelatedItems(self):
         related = viewlet.related_items()
         self.assertEqual([x.Title for x in related], ['Document 2', 'Document 3'])
 
+    def testDeletedRelatedItems(self):
+        # Deleted related items should not cause problems.
+        self.folder._delObject('doc2')
+        request = self.app.REQUEST
+        viewlet = ContentRelatedItems(self.folder.doc1, request, None, None)
+        viewlet.update()
+        related = viewlet.related_items()
+        self.assertEqual([x.Title for x in related], ['Document 3'])
+
 
 class TestDexterityRelatedItemsViewlet(ViewletsTestCase):
 
@@ -174,6 +183,15 @@ def testDexterityFolderRelatedItems(self):
         related = viewlet.related_items()
         self.assertEqual(len(related), 1)
 
+    def testDexterityDeletedRelatedItems(self):
+        # Deleted related items should not cause problems.
+        self.folder._delObject('doc1')
+        request = self.app.REQUEST
+        viewlet = ContentRelatedItems(self.folder.dex1, request, None, None)
+        viewlet.update()
+        related = viewlet.related_items()
+        self.assertEqual([x.id for x in related], ['doc2'])
+
 
 def test_suite():
     from unittest import defaultTestLoader


Repository: plone.app.layout


Branch: refs/heads/2.3.x
Date: 2017-01-13T20:41:23+01:00
Author: Gil Forcada Codinachs (gforcada) <gil.gnome@gmail.com>
Commit: https://github.com/plone/plone.app.layout/commit/3f9516af97997821b658ab3f64d5050e3a8089d6

Merge pull request #114 from plone/fix-deleted-related-dexterity-item-23

Fix error in viewlet when related dexterity item has been deleted. [2.3.x]

Files changed:
M CHANGES.rst
M plone/app/layout/viewlets/content.py
M plone/app/layout/viewlets/tests/test_content.py

diff --git a/CHANGES.rst b/CHANGES.rst
index a53be63..52f8f9b 100644
--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -14,7 +14,8 @@ New features:
 
 Bug fixes:
 
-- *add item here*
+- Fix error in viewlet when related dexterity item has been deleted.
+  [maurits]
 
 
 2.3.16 (2016-12-19)
diff --git a/plone/app/layout/viewlets/content.py b/plone/app/layout/viewlets/content.py
index 83b8c92..7892997 100644
--- a/plone/app/layout/viewlets/content.py
+++ b/plone/app/layout/viewlets/content.py
@@ -176,6 +176,10 @@ def related2brains(self, related):
         brains = []
         for r in related:
             path = r.to_path
+            if path is None:
+                # Item was deleted.  The related item should have been cleaned
+                # up, but apparently this does not happen.
+                continue
             # the query will return an empty list if the user has no
             # permission to see the target object
             brains.extend(catalog(path=dict(query=path, depth=0)))
diff --git a/plone/app/layout/viewlets/tests/test_content.py b/plone/app/layout/viewlets/tests/test_content.py
index 3bf71bc..04b3f58 100644
--- a/plone/app/layout/viewlets/tests/test_content.py
+++ b/plone/app/layout/viewlets/tests/test_content.py
@@ -106,6 +106,15 @@ def testRelatedItems(self):
         related = viewlet.related_items()
         self.assertEqual([x.Title for x in related], ['Document 2', 'Document 3'])
 
+    def testDeletedRelatedItems(self):
+        # Deleted related items should not cause problems.
+        self.folder._delObject('doc2')
+        request = self.app.REQUEST
+        viewlet = ContentRelatedItems(self.folder.doc1, request, None, None)
+        viewlet.update()
+        related = viewlet.related_items()
+        self.assertEqual([x.Title for x in related], ['Document 3'])
+
 
 class TestDexterityRelatedItemsViewlet(ViewletsTestCase):
 
@@ -174,6 +183,15 @@ def testDexterityFolderRelatedItems(self):
         related = viewlet.related_items()
         self.assertEqual(len(related), 1)
 
+    def testDexterityDeletedRelatedItems(self):
+        # Deleted related items should not cause problems.
+        self.folder._delObject('doc1')
+        request = self.app.REQUEST
+        viewlet = ContentRelatedItems(self.folder.dex1, request, None, None)
+        viewlet.update()
+        related = viewlet.related_items()
+        self.assertEqual([x.id for x in related], ['doc2'])
+
 
 def test_suite():
     from unittest import defaultTestLoader


